### ISSUE TITLE
Memory Leak?

### ISSUE THREAD

**Issue by sschepens** [phstc/shoryuken#88]
Memory seems to grow slowly but without limit when using shoryuken with rails environment, a single worker and waiting on an empty queue.

Any ideas what could be causing this behaviour?

**Comment by phstc** [comment]
hey @sschepens which version of Shoryuken are you using? Old versions of Shoryuken had a problem with `message_attributes`, but @elsurudo fixed that https://github.com/phstc/shoryuken/commit/d0e69dc54eaebbd0f3ebe1622577857aa14a0b90

**Comment by sschepens** [comment]
@phstc I'm using latest version

**Comment by leemhenson** [comment]
ruby -v?

**Comment by sschepens** [comment]
ruby 2.2.0p0 (2014-12-25 revision 49005) [x86_64-linux-gnu]

**Comment by leemhenson** [comment]
What sort of growth are you talking about? Can you give numbers / new relic graphs etc?

**Comment by sschepens** [comment]
i'll leave a process running for a few hours recording heap metrics and will come back with that info

**Comment by phstc** [comment]
Any chance it's your worker itself causing the memory leak? Could you try to run Shoryuken with an empty `perform`?

Some questions for troubleshooting:

* How many Shoryuken processes do you have?
* What's the concurrency configured for each?
* What are your queues priorities?
* What's your host/container/whatever CPU (cores) and memory?

**Comment by sschepens** [comment]
Only one process with 5 concurrency, only one queue with a priority of 2.
This behaviour happens on my laptop and in an EC2 instance, the process starts with around 150-160Mb of memory according to htop, and on an EC2 instance the process has been running for a day and it now consumes 700Mb. This instance hast 2000Mb of ram and 1 core.
On my laptop (16Gb ram and 4 cores) after an hour it now consumes 200Mb and continues to grow but very slowly. I have recorded heap metrics every two minutes on my laptop but cant seem to find something that ould indicate the growth of 40 mb.

I dont think running with an empty perform will do much, as i'm hanging on an empty queue.

**Comment by leemhenson** [comment]
What happens if you stick in an occasional ```GC.start```?

**Comment by sschepens** [comment]
That is exactly what i was doing now, and recording metrics before and after gc start

**Comment by phstc** [comment]
@sschepens sorry I missed that part:

> waiting on an empty queue.

You are right, no point on testing an empty work.

I will try to run the same test here to see how it goes.

**Comment by phstc** [comment]
I kept it running for a few hours with default options (checking an empty queue) and the real memory size was consistent around 52MB.

```sh
$ dotenv bin/shoryuken -q default -r ./examples/default_worker.rb -p shoryuken.pid -l shoryuken.log -d

$ top -pid $(cat shoryuken.pid)
```

@sschepens are you using the default `delay`?

My ruby version `ruby 2.0.0p451`. I will bump my ruby to try it with 2.2.

**Comment by sschepens** [comment]
Yes, i'm using default delay.
The symptom i'm having is that on weekends (when we don't deploy and we don't do processing) instances die because shotyuken uses up all memory.
I have two different shoryuken processes on two separate machines and both die.
The things they have in common:
- Same ruby version
- Both include Rails (but only one actually uses rails stack (ActiveRecord) on process.)
- Both use aws-sdk on process
- Both use auto-delete and body_parser with MultiJson + Oj

I still have to record some useful heap information to get a clue of what's going on.

**Comment by phstc** [comment]
hm I will try to run that with Rails, because running with Ruby 2.2.1 the real memory size was consistent around 30MB (better than with 2.0)

**Comment by sschepens** [comment]
Ok, i'm using ruby 2.2.0 and it seems to have a memory leak on symbol gc: https://bugs.ruby-lang.org/issues/10686
Gonna try updating to 2.2.1 to see if it gets fixed

**Comment by curtislinden** [comment]
We are also experiencing memory run-out problems

We are running a worker process in docker with a memory cap of 500MB.

We have filled our queue with fake data that can be partially processed and will generate an exception after a few lines of code. This keeps the worker busy and makes it easier to monitor it's memory usage behavior.

After 12 hours the process hits the memory cap and is shutdown.

We are running ruby mri 2.1.5 and Rails 4.1 environment with Shoryuken 1.0.1

**Comment by phstc** [comment]
>  I'm using ruby 2.2.0 and it seems to have a memory leak on symbol gc: https://bugs.ruby-lang.org/issues/10686

@curtislinden could you try ruby 2.2.1?

**Comment by sschepens** [comment]
I still have leaks on ruby 2.2.1 and 2.2.2 :(
Didnt have time to get good traces of where the memory is leaking

**Comment by curtislinden** [comment]
Same issues with 2.2.2

I found a similar conversation in celluloid and have some insight to share.

I believe this will affect shoryuken as I believe in particular when an exception is caught and the celluloid actor is killed it doesn't have it's memory cleaned up. It may be true for actors that complete a task successfully - as-well 


---
https://github.com/celluloid/celluloid/issues/463#issuecomment-93838990

@curtislinden try to delete actor instance variables in the celluloid finalizer. The "celluloized" object is known to be never collected by GC.
You may check a w/a for DCell: niamster/dcell@61d4e0d

**Comment by sschepens** [comment]
The problem i believe is that shoryuken doesn't recreate actors, they are long living and only recreated if an exception happened (and only workers i believe)

**Comment by curtislinden** [comment]
Since Celluloid was written in the spirit of erlang - I assume that "fail-fast" and "let it die" were good patterns to follow with our workers. In that spirit when our workers can't process a job - they throw exceptions for various cases. In practice these exceptions are common but temporary. 

Since we are throwing exceptions often in our current data-set - I suspect given @sschepens  belief about recreation exception handling in shoryuken may be the place to have cleanup.

**Comment by sschepens** [comment]
I know i've seen exceptions being swallowed in shoryuken, for example here: [fetcher](https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/fetcher.rb#L53-L58)
Also, `watchdog` annotation swallows exceptions and only logs them and seems to be used a lot.

Don't know if there are more occurrences of these, and i'm not really sure if those are causing issues.
Maybe shoryuken should just let exceptions kill all actors and use `supervise` or `Celluloid::SupervisionGroup` to autostart actors.

**Comment by sschepens** [comment]
@phstc @curtislinden @leemhenson 
Just wanting to let you guys I created a new [project](https://github.com/sschepens/poller) I took some code from shoryuken but simplified the architecture. It's not forked because some things are just a complete rewrite.
I saw shoryuken as too complex for what it does or should do (in my opinion :)) and was trying to hunt for possible memory leaks, so I made a few simplifications:
- I removed priority based loadbalancing
- Spawn one long polling fetcher per queue
- Fetchers get messages according to amount of idle processors and waits until at least one is available to get more messages.
- Added configurable pool of processors per queue, which should give a desirable amount of loadbalancing.
- Added a new actor Scheduler which is spawned for each processor now, it's task it's just to take care of the timers of the processor (when as processor is processing it's most definitely doing blocking calls so the processor can't trigger timers while it's processing, this breaks auto_visibility_timeout). I wanted to go for one unique Scheduler but Amazon SDK uses Net::HTTP which is blocking so multiple timers would not trigger concurrently leaving a good chance of breaking auto_visibility_timeout.
- Usage of Actor Registry instead of passing actor on initialization
- Extensive usage of linking and supervising, every actor is restarted when an exception is raised.
- Actors free instance variables on shutdown to prevent memory leaks.

I've also removed some code which i didn't need right now, such as ActiveJob integration.

Anyway, I have been testing and my processes have now been running for a week more or less and I have not had any memory issues.
The project is not documented at all right now and has no tests, but just wanting to let you know that it existed.
Please tell me if you have any issues with me taking code from shoryuken, and if you like it maybe we can work on merging both of them.

Cheers!
Sebastian

**Comment by phstc** [comment]
> I know i've seen exceptions being swallowed in shoryuken, for example here: fetcher
> Also, watchdog annotation swallows exceptions and only logs them and seems to be used a lot.

> Don't know if there are more occurrences of these, and i'm not really sure if those are causing issues.
> Maybe shoryuken should just let exceptions kill all actors and use supervise or Celluloid::SupervisionGroup to autostart actors.

Fetcher exceptions should be rare. And if a processor dies, it will be handled by [manager/trap_exit](https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/manager.rb#L11).

I totally believe in your and I'm sure this issue is happening for you. But I couldn't reproduce it in my production and development environment.

> Please tell me if you have any issues with me taking code from shoryuken, and if you like it maybe we can work on merging both of them.

:+1: for the initiative! Feel free to copy any code you need! A [did the same](https://github.com/phstc/shoryuken#credits) with Sidekiq. Remember to keep the [LGPLv3 License](https://github.com/phstc/shoryuken/blob/master/LICENSE.txt).

@sschepens @curtislinden as you can reproduce the issue, could you try :pray: 

```
# https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/manager.rb#L78-79
@threads.delete(processor.object_id)
@busy.delete processor
remove_instance_variable processor
```
???

Following https://github.com/niamster/dcell/commit/61d4e0d919ae702df22dc9472aec57f5d7aa32ee?

**Comment by curtislinden** [comment]
Hey! yeah We can try this

On Mon, Apr 27, 2015 at 6:36 PM, Pablo Cantero <notifications@github.com>
wrote:

> I know i've seen exceptions being swallowed in shoryuken, for example
> here: fetcher
> Also, watchdog annotation swallows exceptions and only logs them and seems
> to be used a lot.
>
> Don't know if there are more occurrences of these, and i'm not really sure
> if those are causing issues.
> Maybe shoryuken should just let exceptions kill all actors and use
> supervise or Celluloid::SupervisionGroup to autostart actors.
>
> Fetcher exceptions should be rare. And if a processor dies, it will be
> handled by manager/trap_exit
> <https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/manager.rb#L11>
> .
>
> I totally believe in your and I'm sure this issue is happening for you.
> But I couldn't reproduce it in my production and development environment.
>
> Please tell me if you have any issues with me taking code from shoryuken,
> and if you like it maybe we can work on merging both of them.
>
> [image: :+1:] for the initiative! Feel free to copy any code you need! A did
> the same <https://github.com/phstc/shoryuken#credits> with Sidekiq.
> Remember to keep the LGPLv3 License
> <https://github.com/phstc/shoryuken/blob/master/LICENSE.txt>.
>
> @sschepens <https://github.com/sschepens> @curtislinden
> <https://github.com/curtislinden> as you can reproduce the issue, could
> you try [image: :pray:]
>
> # https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/manager.rb#L78-79
> @threads.delete(processor.object_id)
> @busy.delete processor
> remove_instance_variable processor
>
> ???
>
> Following niamster/dcell@61d4e0d
> <https://github.com/niamster/dcell/commit/61d4e0d919ae702df22dc9472aec57f5d7aa32ee>?
>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/phstc/shoryuken/issues/88#issuecomment-96871350>.
>



-- 
Curtis J Schofield | Senior Systems Analyst | U2U
Skype robotarmymade | Second Life Curtis Linden | C 415.632.6001

Linden Lab | Makers of Shared Creative Spaces

**Comment by phstc** [comment]
hey @curtislinden did you have a chance to try that?

**Comment by curtislinden** [comment]
Hey - Super interested in trying this out - just have to get some time set
aside for this. I'll see If i can get to it today.

On Fri, May 8, 2015 at 5:11 AM, Pablo Cantero <notifications@github.com>
wrote:

> hey @curtislinden <https://github.com/curtislinden> did you have a chance
> to try that?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/phstc/shoryuken/issues/88#issuecomment-100213119>.
>



-- 
Curtis J Schofield | Senior Systems Analyst | U2U
Skype robotarmymade | Second Life Curtis Linden | C 415.632.6001

Linden Lab | Makers of Shared Creative Spaces

**Comment by curtislinden** [comment]
Hi - tried it out and it appears that processor isn't an instance variable..

**Comment by phstc** [comment]
Good point! The processors are elements inside the arrays `@ready` and `@busy` (which are instance variables).

The processors flow:

* The manager creates the processors based on the `concurrency` configuration during its initialization: https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/manager.rb#L21
* Then to process messages, they are moved from `@ready` to `@busy`: https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/manager.rb#L95-96
* Then moved back when the processing is done: https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/manager.rb#L64-69

Based on what I understood on this thread, the processors might be kept in memory [when they die](https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/manager.rb#L78-79), increasing the memory usage. The `remove_instance_variable` was an attempt to make sure that we are removing them out from memory when they die. But we probably need to find a different solution for that.

**Comment by curtislinden** [comment]
Interesting.. I think the hypothesis is indeed that the processors might be
kept in memory after they die.

This is an interesting problem.



On Sun, May 10, 2015 at 3:05 PM, Pablo Cantero <notifications@github.com>
wrote:

> Good point! The processors are elements inside the arrays @ready and @busy
> (which are instance variables).
>
> The processors flow:
>
>    - The manager creates the processors based on the concurrency
>    configuration during its initialization:
>    https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/manager.rb#L21
>    - Then to process messages, they are moved from @ready to @busy:
>    https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/manager.rb#L95-96
>    - Then moved back when the processing is done:
>    https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/manager.rb#L64-69
>
> Based on what I understood on this thread, the processors might be kept in
> memory when they die
> <https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/manager.rb#L78-79>,
> increasing the memory usage. The remove_instance_variable was an attempt
> to make sure that we are removing them out from memory when they die. But
> we probably need to find a different solution for that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/phstc/shoryuken/issues/88#issuecomment-100705726>.
>



-- 
Curtis J Schofield | Senior Systems Analyst | U2U
Skype robotarmymade | Second Life Curtis Linden | C 415.632.6001

Linden Lab | Makers of Shared Creative Spaces

**Comment by esebastian** [comment]
Same problem here. I've tried both nilling `processor' and reuse it once deleted from `@busy` but without success.

What I've noticed is that when `processor` is reused without building a new one with `build_processor`, no new messages are processed once the number of messages matches the `concurrency` configuration (i.e. all processors have been used once), and memory keeps stable.

I know this doesn't discard a problem on the processors part (indeed it could be a new evidence), but could it be possible that there may be a problem on the message parsing and digesting part?

**Comment by phstc** [comment]
I couldn't reproduce it locally, but based on what I understood it's happening when the workers are raising exceptions, right? Do you know what kind of exceptions you workers are raising? So I can try to reproduce the same here.

[Does Shoryuken show a reason in the logs](https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/manager.rb#L76)?

> What I've noticed is that when processor is reused without building a new one with build_processor, no new messages are processed once the number of messages matches the concurrency configuration (i.e. all processors have been used once), and memory keeps stable.

Actually [it stops fetching messages only if there isn't any ready processor](https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/manager.rb#L128-134), if there's at least one, it will try to fetch a message for it.

> I know this doesn't discard a problem on the processors part (indeed it could be a new evidence), but could it be possible that there may be a problem on the message parsing and digesting part?

The way Shoryuken manages died processors is the same way [Sidekiq does](https://github.com/mperham/sidekiq/blob/master/lib/sidekiq/manager.rb#L96-L109) - so or both have the same problem, or the problem is in another place in Shoryuken. It could be also with the aws-sdk, but it's strange as I couldn't reproduce it here. Could you create a sample project that causes this problem?

**Comment by esebastian** [comment]
On the exceptions part, yes, workers are raising exceptions in a steady manner; but custom ones. Perhaps I'm doing it the wrong way, but what I do is to raise an exception for every unprocessable message in order to automatically have them reprocessed later. I mean, if I can't process the message in a particular point of time, I raise an exception in order to enqueue the message again and process it later, doing that until the processing can be carried out.

Is there another approach I should be taking?

I'll try set up that sample project for you.

**Comment by sschepens** [comment]
@esebastian Raising exceptions as a method of flow control is not a good practice in most programming languages and is slower. It's always recommended to not raise exceptions if not necessary.
Using shoryuken you can disable automatic message ack and do it yourself when done processing a message.

**Comment by curtislinden** [comment]
@sschepens : I think there is a difference between designing with
exceptions in mind : http://www.c2.com/cgi/wiki?DesigningWithExceptions -
and using it for control flow -
http://www.c2.com/cgi/wiki?DontUseExceptionsForFlowControl

In this case of queue processing - there is an issue with the data, an
exceptional condition is utilized - and the contract cannot be occurred -
in this case what is akin to a transaction is utilized and the message is
not removed from the queue.

This seems like appropriate design for exceptions. Given that the author
chose to use celluloid - following the pattern of 'let it fail' from erlang
- while not letting the failing process take out the rest of the system -
seems to be appropriate.



On Thu, May 14, 2015 at 8:48 AM, sschepens <notifications@github.com> wrote:

> @esebastian <https://github.com/esebastian> Raising exceptions as a
> method of flow control is not a good practice in most programming languages
> and is slower. It's always recommended to not raise exceptions if not
> necessary.
> Using shoryuken you can disable automatic message ack and do it yourself
> when done processing a message.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/phstc/shoryuken/issues/88#issuecomment-102079503>.
>



-- 
Curtis J Schofield | Senior Systems Analyst | U2U
Skype robotarmymade | Second Life Curtis Linden | C 415.632.6001

Linden Lab | Makers of Shared Creative Spaces

**Comment by phstc** [comment]
But maybe as a test, just to make sure that the `processor_died` is the cause of these memory problems.

@esebastian Could you try:

```ruby
class MyWorker
  include Shoryuken::Worker

  shoryuken_options queue: 'my_queue', auto_delete: false

  def perform(sqs_msg, body)    
    # do something that might raise an exception
    sqs_msg.delete
  rescue
    # if it fails and you don't "ack" `sqs_msg.delete` 
    # the message will return to the queue after its visibility timeout expiration
  end
end
```

**Comment by sschepens** [comment]
@curtislinden I think he wasn't talking that there's something wrong with the data, it just wasn't ready to be processed yet he said, that's sort of control flow, saying hey this isn't ready yet, let's just throw an exception, so my automatic ack is skipped, specially when this is the only thing you're aiming for. It could easily be replaced by disabling automatic ack and an immediate return of perform method whenever something is not right.
I like designing with exceptions in mind, i mostly do, but I don't think this is a nice use case.
Talking about Celluloid, I think it's even more important to try and raise fewer exceptions, Erlang is designed for this, Ruby and Celluloid are not. Each Celludloid actor has a thread of it's own and has a lot of overhead in creating/destroying them, this should definitely be considered.

**Comment by esebastian** [comment]
@phstc: I've tried as you asked and the behavior is the same, it keeps leaking memory. Do you still need that sample project?

@sschepens: I think there's a case for anything, but I see your point. I couldn't have expressed mine better than @curtislinden but I will give yours a try. Thanks for the advice.

**Comment by esebastian** [comment]
Any update on this? Can I be of any help?

**Comment by phstc** [comment]
> Do you still need that sample project?

@esebastian could you create a sample project that causes this memory leak problem? So I can reproduce it locally?

Thanks

**Comment by esebastian** [comment]
@phstc Ok, I'll get into it

**Comment by phstc** [comment]
@esebastian tks a lot :beers:

**Comment by phstc** [comment]
@esebastian could you try the master version again, @gshutler did a change that might help with the memory leak problem.

**Comment by esebastian** [comment]
@phstc I'm afraid that the problem persists. I've had the master version running since past Sunday, with different load configurations, and the leak still exists :(

**Comment by esebastian** [comment]
@phstc I'll post a link for the sample project as soon as possible

**Comment by phstc** [comment]
@esebastian thanks. that will help me to debug it.

**Comment by gshutler** [comment]
I'm starting to suspect we're seeing this too. Going to attempt to find the smallest case that reproduces it. Running Ruby v2.2.2

![screen shot 2015-06-19 at 16 09 30](https://cloud.githubusercontent.com/assets/81344/8256274/e396bc38-169d-11e5-9d72-6cf4eabe7827.png)

**Comment by gshutler** [comment]
I think I've got something, it's confusing as hell, taken me on a long, winding road, but I think I've got something.

I'd done a general skim over the code before due to this issue being open and nothing stood out, so it was time to get scientific.

I added this into `Manager#dispatch` to get some high level figures to track:

```ruby
@iterations ||= 0
@iterations += 1

if @iterations % 5000 == 0
  ObjectSpace.garbage_collect
  logger.warn "#{@iterations},#{Time.now.getutc},#{`ps -o rss -p #{$$}`.strip.split.last.to_i * 1024}"
end
```

I'd also output `ObjectSpace.count_objects` before and after GC at various times during my search, 5000 also increased or decreased along the way.

I started looking at polling an empty queue, regardless of the number of workers this should mean only the manager and fetcher actors are doing anything. Also, this is the only section of Shoryuken that differs significantly from Sidekiq and so was my prime suspect. The growth was still observed so I tried to isolate it further, therefore I changed [`Fetcher#received_messages`](https://github.com/phstc/shoryuken/blob/3d1e044f60b340724347d324f044370578b9d18f/lib/shoryuken/fetcher.rb#L21) to do everything but the request and instead just return an empty array to simulate an empty queue. This sped testing up and I was getting similar looking memory growth per iteration, happy days.

`ObjectSpace.count_objects` wasn't flagging anything weird to me, the number of objects weren't increasing out of control, if at all, the GC seemed to always drag things back to the same level, but the memory used by the process would continue to increase.

I tried many things that seemed like obvious targets and would remove them from the loop. For example not calling `Shoryuken.worker_registry.batch_receive_messages?` not doing anything at all within `Fetcher#receive_messages` other than returning an empty array but nothing seemed to make a difference.

Eventually I realised that `Fetcher#fetch` wouldn't really be doing anything other than calling `@manager.async.dispatch` to complete the loop, so I started moving `@manager.async.dispatch and return` down within `Fetcher#fetch` line-by-line.

I found that introducing:

```ruby
logger.debug "Looking for new messages in '#{queue}'"
```

I didn't believe an innocent looking log message could cause any harm so I repeated the test, but its introduction definitely introduce the problem. Eventually, I removed all the logging from `Fetcher` and it did indeed look like the problem had mostly disappeared. Obviously we wouldn't want to lose the logging so I put it back and tried attacking the logging code directly. However, nothing I tried removed the problem.

Eventually I tried converting the logging to a block call style everywhere as this is used in some places and not all. For example:

```ruby
logger.debug { "Looking for new messages in '#{queue}'" }
```

Now this seemed to all but eliminate the growth as this chart shows:

![memory-growth](https://cloud.githubusercontent.com/assets/81344/8260071/2a0195c4-16b9-11e5-991b-c1bf7169575c.png)

I don't understand this if I'm honest. However, the experiments I've run have me convinced this is the source of the majority of the problem. I wondered if Rails is somehow messing with the logger so it's buffering the log messages or something along those lines but I tried various Rails log flushing things and it didn't work.

So there we go, I think it's caused by the logging messages but I have no idea why. I'm going to do a release with a custom branch to see how it behaves in production and will report back.

**Comment by curtislinden** [comment]
This is a very interesting discovery and effective analysis and graph. What
did you use to make the chart?


On Fri, Jun 19, 2015 at 11:30 AM, Garry Shutler <notifications@github.com>
wrote:

> I think I've got something, it's confusing as hell, taken me on a long,
> winding road, but I think I've got something.
>
> I'd done a general skim over the code before due to this issue being open
> and nothing stood out, so it was time to get scientific.
>
> I added this into Manager#dispatch to get some high level figures to
> track:
>
> @iterations ||= 0@iterations += 1
> if @iterations % 5000 == 0
>   ObjectSpace.garbage_collect
>   logger.warn "#{@iterations},#{Time.now.getutc},#{`ps -o rss -p #{$$}`.strip.split.last.to_i * 1024}"end
>
> I'd also output ObjectSpace.count_objects before and after GC at various
> times during my search, 5000 also increased or decreased along the way.
>
> I started looking at polling an empty queue, regardless of the number of
> workers this should mean only the manager and fetcher actors are doing
> anything. Also, this is the only section of Shoryuken that differs
> significantly from Sidekiq and so was my prime suspect. The growth was
> still observed so I tried to isolate it further, therefore I changed
> Fetcher#received_messages
> <https://github.com/phstc/shoryuken/blob/3d1e044f60b340724347d324f044370578b9d18f/lib/shoryuken/fetcher.rb#L21>
> to do everything but the request and instead just return an empty array to
> simulate an empty queue. This sped testing up and I was getting similar
> looking memory growth per iteration, happy days.
>
> ObjectSpace.count_objects wasn't flagging anything weird to me, the
> number of objects weren't increasing out of control, if at all, the GC
> seemed to always drag things back to the same level, but the memory used by
> the process would continue to increase.
>
> I tried many things that seemed like obvious targets and would remove them
> from the loop. For example not calling
> Shoryuken.worker_registry.batch_receive_messages? not doing anything at
> all within Fetcher#receive_messages other than returning an empty array
> but nothing seemed to make a difference.
>
> Eventually I realised that Fetcher#fetch wouldn't really be doing
> anything other than calling @manager.async.dispatch to complete the loop,
> so I started moving @manager.async.dispatch and return down within
> Fetcher#fetch line-by-line.
>
> I found that introducing:
>
> logger.debug "Looking for new messages in '#{queue}'"
>
> I didn't believe an innocent looking log message could cause any harm so I
> repeated the test, but its introduction definitely introduce the problem.
> Eventually, I removed all the logging from Fetcher and it did indeed look
> like the problem had mostly disappeared. Obviously we wouldn't want to lose
> the logging so I put it back and tried attacking the logging code directly.
> However, nothing I tried removed the problem.
>
> Eventually I tried converting the logging to a block call style everywhere
> as this is used in some places and not all. For example:
>
> logger.debug { "Looking for new messages in '#{queue}'" }
>
> Now this seemed to all but eliminate the growth as this chart shows:
>
> [image: memory-growth]
> <https://cloud.githubusercontent.com/assets/81344/8260071/2a0195c4-16b9-11e5-991b-c1bf7169575c.png>
>
> I don't understand this if I'm honest. However, the experiments I've run
> have me convinced this is the source of the majority of the problem. I
> wondered if Rails is somehow messing with the logger so it's buffering the
> log messages or something along those lines but I tried various Rails log
> flushing things and it didn't work.
>
> So there we go, I think it's caused by the logging messages but I have no
> idea why. I'm going to do a release with a custom branch to see how it
> behaves in production and will report back.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/phstc/shoryuken/issues/88#issuecomment-113598856>.
>



-- 
Curtis J Schofield | Senior Systems Analyst | U2U
Skype robotarmymade | Second Life Curtis Linden | C 415.632.6001

Linden Lab | Makers of Shared Creative Spaces

**Comment by gshutler** [comment]
The log messages I added are comma separated values, I just dropped them into Excel to make the chart.

**Comment by curtislinden** [comment]
pragmatic and effective!

On Fri, Jun 19, 2015 at 11:34 AM, Garry Shutler <notifications@github.com>
wrote:

> The log messages I added are comma separated values, I just dropped them
> into Excel to make the chart.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/phstc/shoryuken/issues/88#issuecomment-113599943>.
>



-- 
Curtis J Schofield | Senior Systems Analyst | U2U
Skype robotarmymade | Second Life Curtis Linden | C 415.632.6001

Linden Lab | Makers of Shared Creative Spaces

**Comment by gshutler** [comment]
It seems to have made an improvement, though I don't think it has resolved the problem entirely.

![screen_shot_2015-06-20_at_06_57_56](https://cloud.githubusercontent.com/assets/81344/8266239/c560094a-171a-11e5-9afb-5a5ec9d7d1af.png)

It took around 3 hours to reach a stable level of memory usage rather than 1 with the logging changes contained in the above commit. That stable level of memory is also higher than it was before.

As memory usage does still increase almost linearly over that time I'm not convinced this is entirely solved. However, it definitely seems better than it was before and so I'll open a PR with the logging changes I made.

**Comment by gshutler** [comment]
I probably won't get chance to look at this again until Monday.

There was still a small increase in memory usage in my experiments when using block-style logging so I'll perhaps still be able to investigate this without having to have SQS in the loop.

**Comment by phstc** [comment]
@gshutler :clap: :beers:

**Comment by gshutler** [comment]
Definitely been observing memory leak style behaviour over the weekend. Back on the investigation today as I need to get it sorted before using Shoryuken for all our background tasks!

**Comment by gshutler** [comment]
Hey, I thought I'd update with my progress, unfortunately it's bad news.

I spent over a week looking into the problems, benchmarking, running against our custom branch with various tweaks, GC tuning, etc, to no avail. I also tried switching to Sidekiq (ActiveJob made this super simple) and still observed the same memory growth.

I did come across aws/aws-sdk-ruby#785 which may help some people running on Ruby 2.2.2 as there's a [memory leak in StringIO](https://bugs.ruby-lang.org/issues/10942). I also tried various JSON and XML parsers as some of those aren't publicised as thread safe but nothing seemed to help significantly. 

I'm resigned to the fact that our application, or more likely a gem we're using, is not thread safe and therefore a thread-based processor isn't going to work for us. We've been using Resque for quite a while without any problems so I'm working on a pre-forking SQS message processor which at a conceptual level will work similarly to Shoryuken. In fact I'll likely make it at least somewhat Shoryuken-compatible to make our transition easier. I'm hoping to make that open source once we're happy it's stable.

There are a few tweaks I made to Shoryuken along the way that I think are general improvements so I'll be opening PRs for them once I get time. The main ones that spring to mind are [calling `Aws.eager_load!` is recommended for threaded code in Ruby 1.9 if you support that](https://github.com/gshutler/shoryuken/blob/a79ae9271ce7073ef3c8940d370efb95b01212be/lib/shoryuken/environment_loader.rb#L48), and [caching visibility timeout on queue](https://github.com/gshutler/shoryuken/blob/a79ae9271ce7073ef3c8940d370efb95b01212be/lib/shoryuken/queue.rb#L12) seems a good idea as otherwise you have [an additional SQS request for every message](https://github.com/gshutler/shoryuken/blob/a79ae9271ce7073ef3c8940d370efb95b01212be/lib/shoryuken/middleware/server/timing.rb#L18) for something that will rarely change.

**Comment by sschepens** [comment]
@gshutler I have developed a shoryuken inspired gem, [toiler](https://github.com/sschepens/toiler), that has a simpler aproach and now uses concurrent-ruby for actors and concurrency features, and i have not seen any issue with memory growth.
It's not absolutely compatible and does not support all features just know, check it out and see if it's any use to use, pull requests are also welcome for new functionallity.
It uses long-polling by default which means less SQS requests and caches visibility timeout as well.
Cheers!

**Comment by Senjai** [comment]
@phstc Has this been resolved?

**Comment by phstc** [comment]
@Senjai Not resolved yet. But this is tricky, for example I couldn't reproduce it. Did you? I closed it because of the lack of activity on this thread, but maybe we can re-open it, you have thoughts on it?

**Comment by sschepens** [comment]
I believe sidekiq is in the process of or has already migrated to concurrent-ruby instead of celluloid, because it was having similar issues, rails 5 is also gonna depend on concurrent-ruby.
I did a small sqs consumer proyect using concurrent-ruby and found no issues whatsoever.
@phstc maybe you should consider dropping celluloid.

**Comment by Senjai** [comment]
@phstc We've ran into the issue. We dont care if heroku kills our dyno due to it really as our system is designed around that possibility, so its rather low priority. But it still appears to be a problem. :+1: for dropping celluloid though, but thats a larger discussion

**Comment by twizzzlers** [comment]
Any progress on this? I currently have the same issue, shoryuken gets 60% of memory usage only by listening for new messages in SQS...

**Comment by phstc** [comment]
@twizzzlers there's a big chance it's more related to what you are doing in your worker instead of in Shoryuken itself. What are you processing?

**Comment by twizzzlers** [comment]
I'm executing some jobs with normal database functions, it could be heavy while executing. But my question is why the memory so high even when its only listening for messages?

**Comment by Senjai** [comment]
Its just shoryuken, you can get the same issue with a worker that does nothing

**Comment by twizzzlers** [comment]
So what is my option here? Is there a workaround to this? Should I change to sidekiq or something else?

**Comment by phstc** [comment]
@twizzzlers you can try Sidekiq, it's using [concurrent-ruby](https://github.com/ruby-concurrency/concurrent-ruby), please if you do, let me know the results. I couldn't easily reproduce the memory issue, but that's happening for sure for some people. Long time ago, when I was using Sidekiq (Celluloid implementation), as I use to auto restart Sidekiq using monit service checks, but my workers were doing some (headless) browser operations with libraries known for leaking memory.

**Comment by Senjai** [comment]
Either way, its an issueand until it can be nailed down to a particular library or code path the issue should be open.

**Comment by kwokster10** [comment]
I agree with @Senjai. Since this is an outstanding issue, it should remain open regardless of the amount of activity. We are seeing this in our app but I did not realize the gem had a memory leak. Our process has been dying in the perform method due to some data missing in the body that we were expecting. 

Error on Heroku: `ERROR: Process died, reason: undefined method `[]' for nil:NilClass`. We are correcting this now but am wondering what we can do to prevent this memory leak.

It's causing our memory quota to reach 1GB within 6 hours. See image below.

![shoryuken-memory-graph](https://cloud.githubusercontent.com/assets/11097983/20397947/7ba9a7a8-acb9-11e6-8792-72d6e065f807.png)

**Comment by phstc** [comment]
@Senjai @kwokster10 I don't mind on reopening it, but unfortunately, I can't work on this at moment. Not sure if we can do much with Celluloid, I guess the fix would be to migrate to concurrent-ruby as the other cool guys.

**Comment by twizzzlers** [comment]
We are staying with this issue here in my company for now while we study other options, altough we'd like to continue with shoryuken since it works so much better with AWS SQS...

**Comment by phstc** [comment]
Hey!

I started to work on the `concurrent-ruby` migration 🍻

**Comment by mariokostelac** [comment]
@Senjai @twizzzlers @kwokster10 are you using any middleware? From shoryuken or any other repo?

**Comment by mariokostelac** [comment]
Does not matter. I found an issue. It was not Celluloid, it was not Ruby (or at least not just Celluloid and not just Ruby), it was the way we handled bookkeeping of threads objects.

**Comment by phstc** [comment]
@mariokostelac amazing work on this 👏 

Shoryuken 2.1.2 is out with your fix. For more details: https://github.com/phstc/shoryuken/blob/master/CHANGELOG.md

**Comment by kwokster10** [comment]
@mariokostelac I don't believe we are using any middleware, only our custom `WorkerClass`.

**Comment by mariokostelac** [comment]
@kwokster10 does not matter. We've explained what was the problem and how we've fixed it. Thanks for coming back.

**Comment by phstc** [comment]
Hey all,

Shoryuken v3 is out 🎉 using concurrent-ruby instead of Celulloid, it should fix this leaking issue. Based on [this](https://github.com/phstc/shoryuken/issues/324#issuecomment-283563590) (thanks @paul) the overall memory footprint is lower and more consistent.

Besides that, I added some cool/handy SQS commands, such as: `ls` to list queues and messages (available/in flight), `mv` to move messages from one queue to another, `dump` to dump messages, and `requeue` to enqueue messages from a dump file. Have a look at `shoryuken help sqs` for some guidance.

Full [CHANGELOG](https://github.com/phstc/shoryuken/blob/master/CHANGELOG.md#v300---2017-03-12).

**Comment by elsurudo** [comment]
That's great! Will have to try this one out. Btw @phstc, does Shoryuken use semver? ie. when you incremented the major version, does that mean it's not backwards-compatible?

**Comment by phstc** [comment]
Hi @elsurudo 

Shoryuken uses semver. The version 3.0.0, introduces some incompatible API changes depending on how you use Shoryuken. For instance, [deprecation warnings](https://github.com/phstc/shoryuken/pull/326) and [setting up `AwsConfig` through Shoryuken](https://github.com/phstc/shoryuken/pull/317) were removed. For most projects, I believe people can just bump Shoryuken.

**Comment by phstc** [comment]
Hi all

Just released a new version of Shoryuken 3.1.0 with a new `concurrent-ruby` implementation using Futures and Promises.

Shoryuken now also supports [processing groups](https://github.com/phstc/shoryuken/wiki/Processing-Groups) (set of queues), which means it supports concurrency, exclusive fetcher, long polling etc per group 🤘
